### PR TITLE
chg:dev:No issue:Create ParserFactory to get DDL/SqlQuery Parser

### PR DESCRIPTION
**Pull request by adeshr**
ParserFactory will return the DDL/SqlQuery parser based on the sql query. The factory will store SqlQueryParser state (for caching purpose) and creates new instance only if connection info changes.
Also, added attribute `type` in ParserResult to distinguish between DDL/SelectQuery.

### CODE REVIEW

File: optimizer/src/main/java/com/qubole/quark/planner/parser/ParserFactory.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      try {
+        sqlQueryParser = new SqlQueryParser(info);
+      } catch (QuarkException e) {
+        throw new SQLException(e.getMessage(), e);
+      }
+    }
+    return sqlQueryParser;
+  }
+
+  public Parser getParser(String sql, Properties info, boolean isDirty) throws SQLException {
```
Comment (amoghmargoor) [inline optimizer/src/main/java/com/qubole/quark/planner/parser/ParserFactory.java]: Looks like factory is caching the SqlQueryParser .... Can you change name 'isDirty' to something that tells user of this function about caching aspect ?

File: optimizer/src/main/java/com/qubole/quark/planner/parser/DDLParser.java [hunk 2]
```diff
@@ -34,7 +34,7 @@ public DDLParserResult parse(String sql) {
   public class DDLParserResult extends ParserResult {
     public DDLParserResult(String parsedSql, SqlKind kind,
         RelNode relNode, boolean parseResult) {
-      super(parsedSql, kind, relNode, parseResult);
+      super(parsedSql, kind, relNode, parseResult, "ddl");
```
Comment (amoghmargoor) [inline optimizer/src/main/java/com/qubole/quark/planner/parser/DDLParser.java]: Doesn't SQL Kind tells you about DDL or non DDL ?

**Comment by vrajat** [comment]
You should create an issue for the PR.

**Comment by vrajat** [comment]
Other than the point that @amoghmargoor brought up, it looks good to me. What is the motivation ? You can use ParserFactory in other places such as the Quark Service with QDS ?

File: fatjdbc/src/main/java/com/qubole/quark/fatjdbc/QuarkConnectionImpl.java [hunk 3]
```diff
@@ -74,11 +67,7 @@ protected QuarkConnectionImpl(QuarkDriver driver, AvaticaFactory factory, String
     this.properties.put(InternalProperty.QUOTED_CASING, cfg.quotedCasing());
     this.properties.put(InternalProperty.QUOTING, cfg.quoting());
 
-    try {
-      sqlQueryParser = new SqlQueryParser(info);
-    } catch (QuarkException e) {
-      throw new SQLException(e);
-    }
+    getSqlQueryParser();
```
Comment (vrajat) [inline fatjdbc/src/main/java/com/qubole/quark/fatjdbc/QuarkConnectionImpl.java]: I suggest you remove this one line function.

**Comment by adeshr** [comment]
@vrajat yes. I am planning to use ParserFactory in QuarkService with QDS.

**Comment by vrajat** [comment]
OK

File: fatjdbc/src/main/java/com/qubole/quark/fatjdbc/executor/DDLPlanExecutor.java [hunk 4]
```diff
@@ -98,27 +98,27 @@ public QuarkMetaResultSet execute(ParserResult result)
     }
     if (sqlNode instanceof SqlCreateQuarkDataSource) {
       int id = executeCreateDataSource((SqlCreateQuarkDataSource) sqlNode);
-      connection.resetSelectParser();
+      connection.setIsDirty();
```
Comment (vrajat) [inline fatjdbc/src/main/java/com/qubole/quark/fatjdbc/executor/DDLPlanExecutor.java]: Will this class also be used by Quark Service ? Then that implies:

1. isDirty flag should be moved into `ParserFactory`.
2. DDLPlanExecutor should be using `parserFactory` instead of `connection`.

Comment (adeshr) [inline fatjdbc/src/main/java/com/qubole/quark/fatjdbc/executor/DDLPlanExecutor.java]: As discussed offline, de-coupling of DDLPlanExecutor and Connection will be done in separate PR.
